ble-shelly-btn.js

```diff
@@ -0,0 +1,221 @@
+let ACTIVE_SCAN =
+  typeof CONFIG.shelly_blu_name_prefix !== "undefined" &&
+  CONFIG.shelly_blu_name_prefix !== null;
+
+let uint8 = 0;
+let int8 = 1;
+let uint16 = 2;
+let int16 = 3;
+let uint24 = 4;
+let int24 = 5;
+
+function getByteSize(type) {
+  if (type === uint8 || type === int8) return 1;
+  if (type === uint16 || type === int16) return 2;
+  if (type === uint24 || type === int24) return 3;
+}
+
+let BTH = [];
+BTH[0x00] = { n: "pid", t: uint8 };
+BTH[0x01] = { n: "Battery", t: uint8, u: "%" };
+BTH[0x3a] = { n: "Button", t: uint8 };
+BTH[0x1a] = { n: "Door", t: uint8 };
+BTH[0x2d] = { n: "Window", t: uint8 };
+BTH[0x05] = { n: "Illuminance", t: uint24, f: 0.01 };
+BTH[0x3f] = { n: "Rotation", t: int16, f: 0.1 };
+
+//TODO: Handle 24 bit numbers
+let BTHomeDecoder = {
+  buffer: null,
+  setBuffer: function (buffer) {
+    this.buffer = buffer;
+  },
+  utoi: function (num, bitsz) {
+    let mask = 1 << (bitsz - 1);
+    return num & mask ? num - (1 << bitsz) : num;
+  },
+  getUInt8: function () {
+    return this.buffer.at(0);
+  },
+  getInt8: function () {
+    return this.utoi(this.getUInt8(), 8);
+  },
+  getUInt16LE: function () {
+    return 0xffff & ((this.buffer.at(1) << 8) | this.buffer.at(0));
+  },
+  getInt16LE: function () {
+    return this.utoi(this.getUInt16LE(), 16);
+  },
+  getUInt24LE: function () {
+    return (
+      0x00ffffff &
+      ((this.buffer.at(2) << 16) | (this.buffer.at(1) << 8) | this.buffer.at(0))
+    );
+  },
+  getInt24LE: function () {
+    return this.utoi(this.getUInt24LE(), 24);
+  },
+  getBufValue: function (type) {
+    if (type === uint8) return this.getUInt8();
+    if (type === int8) return this.getInt8();
+    if (type === uint16) return this.getUInt16LE();
+    if (type === int16) return this.getInt16LE();
+    if (type === uint24) return this.getUInt24LE();
+    if (type === int24) return this.getInt24LE();
+    return null;
+  },
+  unpack: function () {
+    if (this.buffer === null) return null;
+    let result = {};
+    let dib = this.buffer.at(0);
+    result["encryption"] = dib & 0x1 ? true : false;
+    result["BTHome_version"] = dib >> 5;
+    if (result["BTHome_version"] !== 2) return null;
+    //Can not handle encrypted data
+    if (result["encryption"] === 1) return null;
+    this.buffer = this.buffer.slice(1);
+
+    while (this.buffer.length > 0) {
+      let _bth = BTH[this.buffer.at(0)];
+      if (_bth === "undefined") return null;
+      this.buffer = this.buffer.slice(1);
+      let _value = this.getBufValue(_bth.t);
+      if (_value === null) return null;
+      if (typeof _bth.f !== "undefined") _value = _value * _bth.f;
+      result[_bth.n] = _value;
+      this.buffer = this.buffer.slice(getByteSize(_bth.t));
+    }
+    return result;
+  },
+};
+
+let ShellyBLUParser = {
+  getData: function (res) {
+    BTHomeDecoder.setBuffer(res.service_data[BTHOME_SVC_ID_STR]);
+    let result = BTHomeDecoder.unpack();
+    return result;
```

Comment:
`ShellyBLUParser.getData()` accepts the entire advertisement result, so it can pack both the BTHome measurement items and Bluetooth-specific attributes in it's response, that is, `BTHparsed` and `BLEData` can both be bundled in the returned value here.

The way it is now, `ShellyBLUParser.getData()`  adds very little value over `BTHomeDecoder.unpack()`.